src/main/java/org/commcare/suite/model/Text.java

```diff
@@ -349,19 +349,22 @@ public void ensureCacheIsParsed() throws XPathSyntaxException {
      * The query evaluation will be abandoned if disposed.
      */
     public Single<String> getDisposableSingleForEvaluation(EvaluationContext ec) {
-        final EvaluationContext[] abandonableContext =
-                new EvaluationContext[] {ec.spawnWithCleanLifecycle()};
+        final EvaluationContext abandonableContext = ec.spawnWithCleanLifecycle();
 
+        final Thread[] toCancel = new Thread[1];
         return Single.fromCallable(new Callable<String>() {
             @Override
             public String call() throws Exception {
-                return evaluate(abandonableContext[0]);
+                toCancel[0] = Thread.currentThread();
+                return evaluate(abandonableContext);
             }
         }).doOnDispose(new io.reactivex.functions.Action() {
             @Override
             public void run() throws Exception {
-                abandonableContext[0].signalAbandoned();
-                abandonableContext[0] = null;
+                if(toCancel[0] != null) {
```

Comment:
@ctsims  The new pattern looks much simpler now 👍  One question, have you been able to test this without calling interrupt explicitly in `doOnDispose()` . I do believe that Rx framework should be interrupting the thread whenever it makes a call to dispose.

Reply:
Hey Shubham!

This is a great question. I meant to make a note that I changed this in the most conservative way I could without validating whether the interrupts happen by default.

I can look into that. I agree that it would be _awesome_ to not need to manually trigger the interrupt.